python/ray/llm/_internal/serve/serving_patterns/data_parallel/dp_group_manager.py

```diff
@@ -0,0 +1,313 @@
+        )
+
+        # Collect all replica IDs to kill (excluding the new registrant)
+        replicas_to_kill = [
+            rid for rid in group.dp_rank_to_replica_id.values() if rid != replica_id
+        ]
+
+        # Clear the group state to prevent cascading kills
+        self._reset_group(group)
+
+        return replicas_to_kill
+
+    async def _kill_replica(self, replica_id: str, group_index: int) -> None:
+        try:
+            actor = ray.get_actor(replica_id)
+            ray.kill(actor, no_restart=False)
+            logger.info(f"Killed replica {replica_id} from group {group_index}")
+        except ValueError:
+            # Actor not found - already dead
+            logger.info(
+                f"Replica {replica_id} from group {group_index} "
+                "already dead, skipping kill"
+            )
+        except Exception as e:
+            logger.warning(
+                f"Failed to kill replica {replica_id} from group " f"{group_index}: {e}"
+            )
+
+    async def _kill_replicas(self, replica_ids: List[str], group_index: int) -> None:
+        """Kill all replicas in the given list."""
+        await asyncio.gather(
+            *[self._kill_replica(rid, group_index) for rid in replica_ids]
+        )
+
+    async def set_dp_master_info(
+        self,
+        group_index: int,
+        dp_address: str,
+        dp_rpc_port: int,
+    ) -> None:
+        """Set the master info for a DP group.
+
+        Called by dp_rank=0 replica after obtaining its address and port.
+        The group must already exist (i.e., register() must be called first).
+
+        Args:
+            group_index: The index of the DP group.
+            dp_address: The IP address of the master.
+            dp_rpc_port: The RPC port of the master.
+        """
+        async with self._group_info_lock:
+            assert group_index in self._group_info, (
+                f"Group {group_index} does not exist. "
+                "register() must be called before set_dp_master_info()."
+            )
+
+            group = self._group_info[group_index]
+            group.master_info = (dp_address, dp_rpc_port)
+
+            # Notify all waiters that the master info is available
+            group.master_info_event.set()
+
+            logger.info(
+                f"Set master info for group {group_index}: "
+                f"address={dp_address}, port={dp_rpc_port}"
+            )
+
+    async def get_dp_master_info(
+        self,
+        group_index: int,
+    ) -> Tuple[str, int]:
+        """Get the master info for a DP group.
+
+        Blocks until the master info is available (set by dp_rank=0). Loops in
+        case the master_info is reset between the time the event is set and we
+        acquire the lock.
+
+        Args:
+            group_index: The index of the DP group.
+
+        Returns:
+            Tuple of (dp_address, dp_rpc_port).
+        """
+        while True:
```

Comment:
why is there a while True here? Doesn't this cause hanging? the blocking nature of this logic is because of await event.wait()?? isn't it?